lib/resty/http_connect.lua

```diff
@@ -160,22 +174,113 @@ local function connect(self, options)
         proxy_port = proxy_uri_t[3]
     end
 
+    local cert_hash
+    -- fallback to non-mTLS when any error
+    repeat
+        if not ssl or not ssl_client_cert or not ssl_client_priv_key then
+            break
+        end
+
+        local cert_type = type(ssl_client_cert)
+        local key_type = type(ssl_client_priv_key)
+
+        if cert_type ~= "cdata" then
+            ngx_log(ngx_WARN, "bad ssl_client_cert: cdata expected, got ", cert_type)
```

Comment:
Why do we only warn? if the client cert is set, it should succeed or error out imho. This could be a security vulnerability, since it automatically lowers security levels. What is the use case for continuing without mtls?

Reply:
This just follows the existing behavior. If `sock:setclientcert` fails, it'll also warn and continue.

```
            ok, err = sock:setclientcert(ssl_client_cert, ssl_client_priv_key)
            if not ok then
              ngx_log(ngx_WARN, "could not set client certificate: ", err)
            end
```

This won't be a security vulnerability. Setting a client cert doesn't necessarily mean the server requires mTLS. It's the server that decides whether the client needs to send a certificate. So if the server doesn't require mTLS then the client cert won't be used even if it's set; if the server does require mTLS but the client doesn't send a valid certificate, the authentication will fail. So I think it makes some sense to warn and continue.

Reply:
That makes sense, though I still think, that if the user sets the cert to begin with, it must be valid. If not, that's an error. So either you don't set it, or you set it to something valid.

Reply:
(it is easier to relax the checks later, than to tighten them)

Reply:
agree. I'll change it to error out directly.